source/common/ssl/context_impl.cc

```diff
@@ -334,13 +350,33 @@ bool ContextImpl::dNSNameMatch(const std::string& dNSName, const char* pattern)
 }
 
 bool ContextImpl::verifyCertificateHashList(
-    X509* cert, const std::vector<std::vector<uint8_t>>& certificate_hash_list) {
+    X509* cert, const std::vector<std::vector<uint8_t>>& expected_hashes) {
   std::vector<uint8_t> computed_hash(SHA256_DIGEST_LENGTH);
   unsigned int n;
   X509_digest(cert, EVP_sha256(), computed_hash.data(), &n);
   RELEASE_ASSERT(n == computed_hash.size());
 
-  for (const auto& expected_hash : certificate_hash_list) {
+  for (const auto& expected_hash : expected_hashes) {
+    if (computed_hash == expected_hash) {
+      return true;
+    }
+  }
+  return false;
+}
+
+bool ContextImpl::verifyCertificateSpkiList(
+    X509* cert, const std::vector<std::vector<uint8_t>>& expected_hashes) {
+  uint8_t* spki = nullptr;
+  const int len = i2d_X509_PUBKEY(X509_get_X509_PUBKEY(cert), &spki);
+  if (len < 0) {
+    return false;
+  }
+
+  std::vector<uint8_t> computed_hash(SHA256_DIGEST_LENGTH);
+  SHA256(spki, len, computed_hash.data());
+  OPENSSL_free(spki);
```

Comment:
no fancy boringssl smart pointer we can use here?

Reply:
Nope, but I change the code to do 2 passes (one to get length, another to write result out - which is the same thing that `i2d_X509_PUBKEY` with `spki` being a pointer to `nullptr` was doing under the hood), and write output to the stack-allocated array, so no need for `OPENSSL_free` anymore.